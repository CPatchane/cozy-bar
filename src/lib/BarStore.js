
/* global fetch */

import { Component } from 'react'
import stack from '../lib/stack'

import CLAUDY_ACTIONS from '../config/claudyActions'

const EXCLUDES = ['settings', 'onboarding']
const CATEGORIES = ['cozy', 'partners', 'ptnb']

export default class BarStore {
  constructor () {
    this.claudyActions = null
    this.installedApps = [] // to cache already fetched apps icons
    this.appsList = [] // all apps, coming soons included
    this.helpLink = ''
    this.settingsAppURL = null
    this.settingsData = {}
  }

  async fetchApps () {
    let apps
    try {
      apps = await Promise.all((await stack.get.apps())
      .filter(app => !EXCLUDES.includes(app.attributes.slug))
      .map(async app => {
        const oldApp = this.installedApps.find(item => item.slug === app.attributes.slug)
        let icon

        if (oldApp && oldApp.icon.cached) {
          icon = oldApp.icon
        } else {
          icon = {
            src: await stack.get.icon(app.links.icon),
            cached: true
          }
        }

        return {
          editor: app.attributes.editor,
          name: app.attributes.name,
          slug: app.attributes.slug,
          href: app.links.related,
          category: CATEGORIES.includes(app.attributes.category) ? app.attributes.category : 'others',
          icon
        }
      }))
      this.installedApps = apps
    } catch (e) {
      apps = [{error: e}]
    }
    return apps
  }

  fetchComingSoonApps () {
    return stack.get.context()
    .then(context => {
      const comingSoonApps = (context.data && context.data.attributes &&
      context.data.attributes['coming_soon'] &&
      Object.values(context.data.attributes['coming_soon'])) || []

      return comingSoonApps.map(app => {
        let icon

        try {
          icon = app.slug && {
            cached: true,
            src: require(`../assets/icons/comingsoon/icon-${app.slug}.svg`)
          }
        } catch (error) {
          console.warn && console.warn(`Cannot retrieve icon for app ${app.name}:`, error.message)
        }

        return Object.assign({}, app, {
          comingSoon: true,
          icon: icon
        })
      })
    })
    .catch(error => {
      console.warn && console.warn(`Cozy-bar cannot fetch comming soon apps: ${error.message}`)
      return []
    })
  }

  async fetchAppsList () {
    const apps = await this.fetchApps()
    const comingSoonApps = await this.fetchComingSoonApps()
    this.appsList = apps.concat(comingSoonApps)
    return this.appsList
  }

  getClaudyActions () {
    if (this.claudyActions) return Promise.resolve(this.claudyActions)
    return stack.get.context()
    .then(context => {
      const contextActions = (context.data && context.data.attributes && context.data.attributes['claudy_actions']) || null
      if (!contextActions) return null
      // get an arrays of action
      const claudyActions = contextActions.map(slug => {
        if (CLAUDY_ACTIONS.hasOwnProperty(slug)) {
          // adding also the action slug
          return Object.assign({}, CLAUDY_ACTIONS[slug], { slug })
        }
      }).filter(action => action)
      this.claudyActions = claudyActions
      return {
        actions: claudyActions
      }
    })
    .catch(error => {
      console.warn && console.warn(`Cozy-bar cannot fetch Claudy: ${error.message}`)
      return null
    })
  }

  getHelpLink () {
    if (this.helpLink) return Promise.resolve(this.helpLink)
    return stack.get.context()
      .then(context => {
        this.helpLink = (context.data && context.data.attributes && context.data.attributes['help_link']) || null
        return this.helpLink
      })
  }

  getStorageStats () {
    return stack.get.storageStats()
      .then(statsObject => statsObject)
      .catch(e => { return {error: e.name} })
  }

  getSettingsAppURL () {
    // If the `settings` app is available, it will used to add the links 'Profile' and 'Connected Devices'
    if (this.settingsAppURL) return Promise.resolve(this.settingsAppURL)
    return stack.get.settingsAppURL()
      .then(settingsAppURL => {
        this.settingsAppURL = settingsAppURL
        return this.settingsAppURL
      })
      .catch(e => {
        console.warn('Settings app is unavailable, settings links are disabled')
        return null
      })
  }

  async fetchSettingsData () {
    this.settingsData = {
      storageStats: await this.getStorageStats(),
      settingsAppURL: await this.getSettingsAppURL(),
      helpLink: await this.getHelpLink()
    }
  }

  logout () {
    stack.logout()
  }
}

export class Provider extends Component {
  getChildContext () {
    return { store: this.store }
  }

  constructor (props, context) {
    super(props, context)
    this.store = props.store
  }

  render ({children}) {
    return (children && children[0]) || null
  }
}
