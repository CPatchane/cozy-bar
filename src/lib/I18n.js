/**
 * preact plugin that provides an I18n helper using a Higher Order Component.
 */

'use strict'

import React, { Component } from 'react'
import Polyglot from 'node-polyglot'

export const DEFAULT_LANG = 'en'

export let _polyglot

const initTranslation = (lang, dictRequire, context, defaultLang = DEFAULT_LANG) => {
  _polyglot = new Polyglot({
    phrases: dictRequire(defaultLang),
    locale: defaultLang
  })

  // Load global locales
  if (lang && lang !== defaultLang) {
    try {
      const dict = dictRequire(lang)
      _polyglot.extend(dict)
      _polyglot.locale(lang)
    } catch (e) {
      console.warn(`The dict phrases for "${lang}" can't be loaded`)
    }
  }

  // Load context locales
  if (context) {
    const dict = dictRequire(lang, context)
    _polyglot.extend(dict)
  }

  return _polyglot
}

// Provider root component
export class I18n extends Component {
  constructor (props) {
    super(props)
    this.init(this.props)
  }

  init (props) {
    const { lang, dictRequire, context, defaultLang } = props

    this.translation = initTranslation(lang, dictRequire, context, defaultLang)
  }

  getChildContext () {
    return {
      t: this.translation.t.bind(this.translation)
    }
  }

  componentWillReceiveProps (newProps) {
    if (newProps.locale !== this.props.locale) {
      this.init(newProps)
    }
  }

  render () {
    return (this.props.children && this.props.children[0]) || null
  }
}

I18n.propTypes = {
  lang: React.PropTypes.string.isRequired,      // current language.
  dictRequire: React.PropTypes.func.isRequired, // A callback to load locales.
  context: React.PropTypes.string,              // current context.
  defaultLang: React.PropTypes.string           // default language. By default is 'en'
}

I18n.childContextTypes = {
  t: React.PropTypes.func
}

// higher order decorator for components that need `t`
export const translate = () => {
  return (WrappedComponent) => {
    const _translate = (props, context) => (
      <WrappedComponent {...props} t={context.t} />
    )
    return _translate
  }
}
