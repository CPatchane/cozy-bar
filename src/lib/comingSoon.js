function addComingSoonApps (stackApps) {
  const comingSoonAppsList = [
    {
      'editor': 'Cozy',
      'name': 'Bank',
      'slug': 'bank',
      'category': 'cozy'
    },
    {
      'editor': 'Cozy',
      'name': 'Health',
      'slug': 'health',
      'category': 'cozy'
    }
  ]

  const comingSoonApps = comingSoonAppsList
    // drop coming soon apps already installed
    .filter((a) => !stackApps.filter(st => st.slug === a.slug).length)
    // consolidate
    .map(a => {
      a.comingSoon = true
      a.l10n = false
      a.icon = {
        cached: true,
        src: require(`../assets/icons/comingSoon/icon-${a.slug}.svg`)
      }
      return a
    })

  // merge to installed apps
  return stackApps.concat(comingSoonApps)
}

export default addComingSoonApps
