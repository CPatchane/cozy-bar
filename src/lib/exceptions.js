class ForbiddenException extends Error {
  constructor (message) {
    super()

    this.name = 'Forbidden'
    this.message = message || 'The application does not have permission to access this resource.'
    this.stack = (new Error()).stack
  }
}

class ServerErrorException extends Error {
  constructor (message) {
    super()

    this.name = 'ServerError'
    this.message = message || 'A server error occurred'
    this.stack = (new Error()).stack
  }
}

class UnavailableStackException extends Error {
  constructor (message) {
    super()

    this.name = 'UnavailableStack'
    this.message = message || 'The stack is temporarily unavailable'
    this.stack = (new Error()).stack
  }
}

class UnauthorizedStackException extends Error {
  constructor (message) {
    super()

    this.name = 'UnauthorizedStack'
    this.message = message || 'The app is not allowed to access to the requested resource'
    this.stack = (new Error()).stack
  }
}

class UnavailableSettingsException extends Error {
  constructor (message) {
    super()

    this.name = 'UnavailableSettings'
    this.message = message || "The 'Settings' application isn't available or installed in the stack"
    this.stack = (new Error()).stack
  }
}

export {
  ForbiddenException,
  ServerErrorException,
  UnavailableStackException,
  UnavailableSettingsException,
  UnauthorizedStackException
}
