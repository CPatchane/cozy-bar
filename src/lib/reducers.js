import { combineReducers } from 'redux'
import * as content from './content'

export default combineReducers({
  content: content.reducer
})

export const getContent = (state, location) => {
  if (!location) {
    throw new Error('Location must provided')
  }
  return content.getContent(state.content, location)
}
