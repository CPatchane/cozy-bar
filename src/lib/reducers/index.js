import { combineReducers } from 'redux'
import * as content from './content'
import * as locale from './locale'
import appsReducer, * as apps from './apps'

const proxy = (attr, method) => {
  return (state, ...args) => {
    return method(state[attr], ...args)
  }
}

const setContent = content.setContent
const setLocale = locale.setLocale
const fetchApps = apps.fetchApps
const setInfos = apps.setInfos
export { setContent, setLocale, fetchApps, setInfos }

export const getContent = proxy('content', content.getContent)
export const getLocale = proxy('locale', locale.getLocale)
export const getApps = proxy('apps', apps.getApps)
export const isAppListForbidden = proxy('apps', apps.isAppListForbidden)
export const getCurrentApp = proxy('apps', apps.getCurrentApp)

export const reducers = {
  content: content.reducer,
  locale: locale.reducer,
  apps: appsReducer
}

export default combineReducers(reducers)
