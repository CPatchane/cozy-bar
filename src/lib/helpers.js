import { t } from '../lib/i18n'

// Take an items array and return an array of category objects with the matching category slug and items
export function getCategorizedItems (items) {
  if (items[0] instanceof Array) return null // doesn't handle this case
  const categorizedItemsObject = items.reduce((accumulator, item) => {
    accumulator[item.category] = accumulator[item.category] || []
    accumulator[item.category].push(item)
    return accumulator
  }, {})

  return Object.keys(categorizedItemsObject)
    .map(category => {
      return {slug: category, items: categorizedItemsObject[category]}
    })
    // categories alphabetical sorting
    .sort((c1, c2) => {
      if (c1.slug === 'others') return 1
      if (c2.slug === 'others') return -1
      if (t(`Categories.${c1.slug}`) > t(`Categories.${c2.slug}`)) return 1
      if (t(`Categories.${c1.slug}`) < t(`Categories.${c2.slug}`)) return -1
      return 0
    })
}
