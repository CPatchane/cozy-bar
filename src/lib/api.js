import React, { Component } from 'react'
import { setContent } from './content'
import { setLocale } from './locale'
import { getContent } from './reducers'

const upperFirstLetter = val => {
  return val[0].toUpperCase() + val.slice(1)
}

/**
 * Wraps argument into a React element if it is a string. Is used
 * for setBar{Left,Right,Center} to be able to pass HTML
 *
 * @param  {[type]} v [description]
 * @return {[type]}   [description]
 */
const wrapInElement = v => {
  if (typeof v === 'string') {
    return <span dangerouslySetInnerHTML={{__html: v}} />
  } else {
    return v
  }
}

/**
 * Creates a React component that enables to access store
 * properties in a declarative way.
 *
 * @param  {BarStore} store
 */
const barContentComponent = (store, location) => class extends Component {
  componentDidMount () {
    this.prev = getContent(store.getState(), location)
    this.setContent(this.props.children)
  }

  setContent (content) {
    store.dispatch(setContent(location, content ? React.Children.only(content) : content))
  }

  componentWillUnmount () {
    this.setContent(this.prev)
  }

  componentDidUpdate (prevProps) {
    if (this.props.children !== prevProps.children) {
      this.setContent(this.props.children)
    }
  }

  render () {
    return null
  }
}

/**
 * Creates a public API
 *
 * - getters/setters for public attributes
 * - React components that act as getters/setters
 *
 * @param  {ReduxStore} store - Store on which the API will act
 * @return {object} - Methods of the public API
 */
export default store => {
  // setBar{Left,Right,Center} and <Bar{Left,Right,Center} />
  const locations = ['left', 'center', 'right']
  const methods = {}
  locations.forEach(location => {
    const upperLoc = upperFirstLetter(location)

    /// expose JS API
    methods[`setBar${upperLoc}`] = value => (
      store.dispatch(setContent(location, wrapInElement(value)))
    )

    // expose React API
    methods[`Bar${upperLoc}`] = barContentComponent(store, location)
  })

  methods.setLocale = lang => store.dispatch(setLocale(lang))

  return methods
}
