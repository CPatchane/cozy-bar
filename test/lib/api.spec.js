import React from 'react'
import { shallow } from 'enzyme'

import createStore from 'lib/store'
import api from 'lib/api'
import { getContent as _getContent } from 'lib/reducers'

const store = createStore()
const {
  setBarLeft,
  setBarRight,
  setBarCenter,
  BarLeft,
  BarRight,
  BarCenter,
  setLocale
} = api(store)

const getContent = location => {
  return _getContent(store.getState(), location)
}

describe('api spec', function () {
  it('should set content of the store via imperative api', function () {
    setBarLeft('content left')
    expect(getContent('left')).toMatchSnapshot()

    setBarRight('<p>content right</p>')
    expect(getContent('right')).toMatchSnapshot()

    setBarCenter('content center')
    expect(getContent('center')).toMatchSnapshot()
  })

  it('should set content of the store via declarative api', function () {
    shallow(<BarLeft><div>Left</div></BarLeft>)
    expect(getContent('left')).toMatchSnapshot()
    shallow(<BarRight>Right</BarRight>)
    expect(getContent('right')).toMatchSnapshot()
    shallow(<BarCenter><span>Center</span></BarCenter>)
    expect(getContent('center')).toMatchSnapshot()
  })

  it('should set locale', function () {
    // default lang is `en`
    expect(store.getState().locale).toBe('en')
    setLocale('fr')
    expect(store.getState().locale).toBe('fr')
    setLocale('de')
    expect(store.getState().locale).toBe('de')
  })
})
