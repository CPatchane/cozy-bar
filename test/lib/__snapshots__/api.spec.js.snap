// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api spec should set content of the store via declarative api 1`] = `
<div>
  Left
</div>
`;

exports[`api spec should set content of the store via declarative api 2`] = `"Right"`;

exports[`api spec should set content of the store via declarative api 3`] = `
<span>
  Center
</span>
`;

exports[`api spec should set content of the store via imperative api 1`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "content left",
    }
  }
/>
`;

exports[`api spec should set content of the store via imperative api 2`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p>content right</p>",
    }
  }
/>
`;

exports[`api spec should set content of the store via imperative api 3`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "content center",
    }
  }
/>
`;
